src/rust/engine/src/python.rs

```diff
@@ -207,6 +207,19 @@ impl Function {
     });
     format!("{module}:{line_no}:{name}")
   }
+
+  /// The function represented as `path.to.module:func_name`. Used to identify a
+  /// rule when called by name.
+  pub fn identifier(&self) -> String {
+    let (module, name) = Python::with_gil(|py| {
+      let val = self.0.to_value();
+      let obj = (*val).as_ref(py);
+      let module: String = externs::getattr(obj, "__module__").unwrap();
+      let name: String = externs::getattr(obj, "__name__").unwrap();
+      (module, name)
+    });
+    format!("{module}:{name}")
+  }
```

Comment:
Going generic might look like having the "production" `RuleId` implementation be something like:
```rust
enum RuleId {
  Native(&'static str),
  Python(crate::python::Function),
}
```

It _might_ be worthwhile to go generic in this PR if it helps to avoid name collisions? But up to you.

Reply:
The name collisions are all on the Python side, and I've fixed them another way. I don't think this would help with that - the collisions are due to dynamic re-registration of the same function, so the `Function` alone is not enough to disambiguate, we need info from the `@rule` decorator.

Reply:
Not sure we will ever need to go generic here, just a newtyped string seems to be enough.